File: src/java/magma/api/collect/head/SingleHead.java
package magma.api.collect.head;

import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;

public final class SingleHead<T> implements Head<T> {
    private final T element;
    private boolean retrieved;

    public SingleHead(T element) {
        this.element = element;
        this.retrieved = false;
    }

    @Override
    public Option<T> next() {
        if (this.retrieved) {
            return new None<T>();
        }

        this.retrieved = true;
        return new Some<T>(this.element);
    }
}

File: src/node/magma/api/collect/head/SingleHead.ts
/*[
	Actual, 
	Characters, 
	Collector, 
	Console, 
	EmptyHead, 
	Files, 
	FlatMapHead, 
	Head, 
	HeadedQuery, 
	Lists, 
	MapHead, 
	Namespace, 
	RangeHead, 
	SingleHead, 
	Strings
]*/
import { Head } from "../../../../magma/api/collect/head/Head";
import { Option } from "../../../../magma/api/option/Option";
import { None } from "../../../../magma/api/option/None";
import { Some } from "../../../../magma/api/option/Some";
export class SingleHead<T> implements Head<T> {
	element: T;
	retrieved: boolean;
	constructor (element: T) {
		this/*auto*/.element = element/*T*/;
		this/*auto*/.retrieved = false/*auto*/;
	}
	next(): Option<T> {
		if (this/*auto*/.retrieved) {
			return new None<T>(/*auto*/);
		}
		this/*auto*/.retrieved = true/*auto*/;
		return new Some<T>(this/*auto*/.element);
	}
}
